File: W11/class-problems/cross.py
"""

    https://cses.fi/problemset/task/2189

"""

def det(p1, p2):
    return p1[0] * p2[1] - p1[1] * p2[0]

def subtract(p1, p2):
    return (p1[0] - p2[0], p1[1] - p2[1])

def main():
    t = int(input())
    for _ in range(t):
        x1, y1, x2, y2, x3, y3 = map(int, input().split())
        p1 = (x1, y1)
        p2 = (x2, y2)
        p3 = (x3, y3)

        cp = det(subtract(p2, p1), subtract(p3, p1))

        if cp == 0:
            print("TOUCH")
        elif cp > 0:
            print("LEFT")
        else:
            print("RIGHT")

if __name__ == "__main__":
    main()

File: W11/class-problems/cross.cpp
#include <iostream>
using namespace std;

#define ll long long

pair<ll,ll> operator-(const pair<ll,ll> &p1, const pair<ll,ll> &p2) {
    return {p1.first - p2.first, p1.second - p2.second};
}

ll det(const pair<ll,ll> &p1, const pair<ll,ll> &p2) {
    return p1.first * p2.second - p1.second*p2.first;
}

int main() {
    int t;
    cin >> t;
    for(int zzz = 0; zzz < t; zzz++) {
        ll x1,y1,x2,y2,x3,y3;
        cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3;
        pair<ll, ll> p1={x1,y1};
        pair<ll, ll> p2={x2,y2};
        pair<ll, ll> p3={x3,y3};
        ll cp = det(p2-p1, p3-p1);
        if(cp == 0) {
            cout << "TOUCH" << endl;
        } else if (cp > 0) {
            cout << "LEFT" << endl;
        } else {
            cout << "RIGHT" << endl;
        }
    }

    return 0;
}